air/src/proof.rs

```diff
@@ -1,102 +1,109 @@
     }
 
+    /// Returns the hash function used during proof generation process.
+    pub const fn hash_fn(&self) -> HashFunction {
+        self.hash_fn
+    }
+
+    /// Returns conjectured security level of this proof in bits.
+    pub fn security_level(&self) -> u32 {
+        match self.hash_fn {
+            HashFunction::Blake3_192 => self.proof.security_level::<Blake3_192>(true),
+            HashFunction::Blake3_256 => self.proof.security_level::<Blake3_192>(true),
+            HashFunction::Rpo256 => self.proof.security_level::<Blake3_192>(true),
+        }
+    }
+
     // SERIALIZATION / DESERIALIZATION
     // --------------------------------------------------------------------------------------------
 
     /// Serializes this proof into a vector of bytes.
     pub fn to_bytes(&self) -> Vec<u8> {
         let mut bytes = self.proof.to_bytes();
-        if bytes.is_empty() {
-            // TODO maybe error?
-            return vec![];
-        }
-        bytes.insert(0, self.hasher as u8);
+        assert!(!bytes.is_empty(), "invalid STARK proof");
+        // TODO: ideally we should write hash function into the proof first to avoid reallocations
+        bytes.insert(0, self.hash_fn as u8);
         bytes
     }
 
     /// Reads the source bytes, parsing a new proof instance.
-    ///
-    /// Will expect the output of `[Self::to_bytes]`.
     pub fn from_bytes(source: &[u8]) -> Result<Self, DeserializationError> {
         if source.len() < 2 {
             return Err(DeserializationError::UnexpectedEOF);
         }
-        let hasher = HashFunction::try_from(source[0])?;
+        let hash_fn = HashFunction::try_from(source[0])?;
         let proof = StarkProof::from_bytes(&source[1..])?;
-        Ok(Self { hasher, proof })
+        Ok(Self::new(proof, hash_fn))
     }
 
     // DESTRUCTOR
     // --------------------------------------------------------------------------------------------
 
-    /// Extracts the underlying hasher and`[StarkProof]`.
-    pub fn into_inner(self) -> (HashFunction, StarkProof) {
-        (self.hasher, self.proof)
+    /// Returns components of this execution proof.
+    pub fn into_parts(self) -> (HashFunction, StarkProof) {
+        (self.hash_fn, self.proof)
     }
 }
 
 // PROOF OPTIONS
 // ================================================================================================
 
-/// A set of arguments to specify how the Miden proofs should be constructed.
+/// A set of parameters specifying how Miden VM execution proofs are to be generated.
 #[derive(Debug, Clone, Eq, PartialEq)]
 pub struct ProofOptions {
-    hasher: HashFunction,
     options: WinterProofOptions,
+    hash_fn: HashFunction,
 }
 
 impl ProofOptions {
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
 
-    /// Creates a new instance with arbitrary hasher.
+    /// Creates a new instance of [ProofOptions] from the specified parameters.
```

Comment:
nit: should be `[ProofOptions]` so rustdocs can link it properly

Reply:
I think that's how it is now - am I missing something?

Reply:
We should have it between `

GH render will suppress it

Reply:
Hmm, I ran `cargo doc --package miden-air --open` and it was able to link correctly as is.

Reply:
True! Must be a patch somewhere in cargo doc. Tested and working here; resolving the comment